src/types.rs

```diff
@@ -330,3 +331,119 @@ impl DestinationSlot {
+    /// the corresponding SQE.
+    pub fn parse(buf: &'buf [u8], name_len: u32, control_len: u32) -> Result<Self, ()> {
+        if buf.len() < std::mem::size_of::<sys::io_uring_recvmsg_out>() {
+            return Err(());
+        }
+
+        let header = sys::io_uring_recvmsg_out {
+            namelen: u32::from_ne_bytes(buf[0..4].try_into().unwrap()),
+            controllen: u32::from_ne_bytes(buf[4..8].try_into().unwrap()),
+            payloadlen: u32::from_ne_bytes(buf[8..12].try_into().unwrap()),
+            flags: u32::from_ne_bytes(buf[12..16].try_into().unwrap()),
+        };
+
+        let total_len = (name_len + control_len + header.payloadlen) as usize;
+        if Self::DATA_START + total_len > buf.len() {
+            return Err(());
+        }
+
+        Ok(Self {
+            buf,
+            header,
+            msghdr_name_len: name_len,
+            msghdr_control_len: control_len,
+        })
+    }
+
+    /// Return the length of the incoming `name` data.
+    ///
+    /// This may be larger than the size of the content returned by
+    /// `name_data()`, if the kernel could not fit all the incoming
+    /// data in the provided buffer size. In that case, name data in
+    /// the result buffer gets truncated.
+    pub fn incoming_name_len(&self) -> u32 {
+        self.header.namelen
+    }
+
+    /// Return whether the incoming name data was larger than the provided limit/buffer.
+    ///
+    /// When `true`, data returned by `name_data()` is truncated and
+    /// incomplete.
+    pub fn is_name_data_truncated(&self) -> bool {
+        self.header.namelen > self.msghdr_name_len
+    }
+
+    /// Message control data, with the same semantics as `msghdr.msg_control`.
+    pub fn name_data(&self) -> &[u8] {
+        let name_start = Self::DATA_START;
+        let name_size = u32::min(self.header.namelen, self.msghdr_name_len);
+        let name_end = name_start + name_size as usize;
+        &self.buf[name_start..name_end]
+    }
+
+    /// Return the length of the incoming `control` data.
+    ///
+    /// This may be larger than the size of the content returned by
+    /// `control_data()`, if the kernel could not fit all the incoming
+    /// data in the provided buffer size. In that case, control data in
+    /// the result buffer gets truncated.
+    pub fn incoming_control_len(&self) -> u32 {
+        self.header.controllen
+    }
+
+    /// Return whether the incoming control data was larger than the provided limit/buffer.
+    ///
+    /// When `true`, data returned by `control_data()` is truncated and
+    /// incomplete.
+    pub fn is_control_data_truncated(&self) -> bool {
+        self.header.controllen > self.msghdr_control_len
+    }
+
+    /// Message control data, with the same semantics as `msghdr.msg_control`.
+    pub fn control_data(&self) -> &[u8] {
+        let control_start = Self::DATA_START + self.msghdr_name_len as usize;
```

Comment:
398L maybe take `header.namelen`, but use `msghdr_name_len` unconditionally here, which is somewhat inconsistent

Reply:
This should be correct as is, because the fixed length of the **name field** is `msghdr_name_len`, while the dynamic length of the **name data** is the minimum between the field length and the length of the actual result data. In this case we want to skip the whole **field**, whose length is fixed upfront regardless of the result.

But I agree this may be confusing. I've added a docstring on both `msghdr_name_len` and `msghdr_control_len` which hopefully could clarify the semantics.